File: c/muse/data_structures/ArrayStackTest.c
#include <muse/data_structures/ArrayStack.h>
#include <muse/util/TestRunner.h>

bool testArrayStack(void) {
    size_t size = 8192;

    ArrayStack *stack = ArrayStack_new(0);

    for (size_t i = 1; i <= size; i++) {
        ArrayStack_push(stack, (int) i);
    }

    ArrayStack_shrink(stack);

    if (ArrayStack_size(stack) != size) {
        return false;
    }

    if (ArrayStack_capacity(stack) != size) {
        return false;
    }

    for (size_t i = size; i > 0; i--) {
        if ((size_t) ArrayStack_peek(stack) != i) {
            return false;
        }
        ArrayStack_pop(stack);
    }

    ArrayStack_shrink(stack);

    if (!ArrayStack_isEmpty(stack)) {
        return false;
    }

    return ArrayStack_capacity(stack) == 0;
}

int main(void) {
    TestRunner_pick(&testArrayStack);
}

File: python/muse/data_structures/array_stack_test.py
from muse.data_structures import tests
from muse.data_structures.array_stack import ArrayStack
from muse.util import test_runner


def test_array_stack() -> bool:
    return tests.stack_derive(ArrayStack)()


def main() -> None:
    test_runner.pick(test_array_stack)


if __name__ == "__main__":
    main()
